# Git's Objects

- git's object storage is
    - key value storage
    - key is SHA-1 hash of object, with some additional information like size

- 4 types of objects are there, they are reffered by
    - branches
    - HEAD

## 4 Types of git objects

- tree (directories)
- blobs (files)
- commits
- tags


In [4]:
pwd

/home/ncoder/02_wos/jupyter-lab/notebooks/03_git/02_git-mastering-version-control-sfrln/2-git-version-control-cookbook


In [5]:
cd ../2-2-git-version-control-cookbook

In [6]:
pwd

/home/ncoder/02_wos/jupyter-lab/notebooks/03_git/02_git-mastering-version-control-sfrln/2-2-git-version-control-cookbook


In [7]:
git clone https://github.com/dvaske/data-model.git

Cloning into 'data-model'...
remote: Enumerating objects: 45, done.
remote: Counting objects: 100% (45/45), done.
remote: Compressing objects: 100% (26/26), done.
remote: Total 45 (delta 11), reused 45 (delta 11), pack-reused 0
Unpacking objects: 100% (45/45), done.


In [8]:
ls

 c1.navigating-git.ipynb  'c2.gits objects.ipynb'   data-model   images


In [11]:
cd data-model;
pwd

bash: cd: data-model: No such file or directory
/home/ncoder/02_wos/jupyter-lab/notebooks/03_git/02_git-mastering-version-control-sfrln/2-2-git-version-control-cookbook/data-model


## Looking at COMMIT object

### HEAD points to
- current snapshot/commit

#### command used to look at a COMMIT

- `git cat-file`

In [12]:
git cat-file -p HEAD

tree 34fa038544bcd9aed660c08320214bafff94150b
parent a90d1906337a6d75f1dc32da647931f932500d83
author Aske Olsson <aske.olsson@switch-gears.dk> 1386933960 +0100
committer Aske Olsson <aske.olsson@switch-gears.dk> 1386941455 +0100

This is the subject line of the commit message

It should be followed by a blank line then the body, which is this text. Here
you can have multiple paragraphs etc. and explain your commit. It's like an
email with subject and body, so get people's attention in the subject


#### git cat-file output

- `cat-file -p` pretty prints given object
    - in this case, HEAD which points to master branch, which inturn points to latest commit on master branch
- what does commit object consists of?
    - root tree (tree 34fa038544bcd9aed660c08320214bafff94150b)
    - parent commit object's id (parent a90d1906337a6d75f1dc32da647931f932500d83)
    - author and timestamp information (author Aske Olsson <aske.olsson@switch-gears.dk> 1386933960 +0100)
    - committer and timestamp information (committer Aske Olsson <aske.olsson@switch-gears.dk> 1386941455 +0100)
    - commit message
    

### differnce between git commit author and committer

The author is the person who originally wrote the code. The committer, on the other hand, is assumed to be the person who committed the code on behalf of the original author. This is important in Git because Git allows you to rewrite history, or apply patches on behalf of another person.

ref: https://stackoverflow.com/questions/18750808/difference-between-author-and-committer-in-git

## Looking into TREE object

**How**
- using `git cat-file -p` command with the tree ID as the argument

In [15]:
git cat-file -p 34fa038544bcd9aed660c08320214bafff94150b

100644 blob f21dc2804e888fee6014d7e5b1ceee533b222c15	README.md
040000 tree abc267d04fb803760b75be7e665d3d69eeed32f8	a_sub_directory
100644 blob b50f80ac4d0a36780f9c0636f43472962154a11a	another-file.txt
100644 blob 92f046f17079aa82c924a9acf28d623fcb6ca727	cat-me.txt
100644 blob bb2fe940924c65b4a1cefcbdbe88c74d39eb23cd	hello_world.c


### git cat-file 

In [16]:
git cat-file -h

usage: git cat-file (-t [--allow-unknown-type] | -s [--allow-unknown-type] | -e | -p | <type> | --textconv | --filters) [--path=<path>] <object>
   or: git cat-file (--batch | --batch-check) [--follow-symlinks] [--textconv | --filters]

<type> can be one of: blob, tree, commit, tag
    -t                    show object type
    -s                    show object size
    -e                    exit with zero when there's no error
    -p                    pretty-print object's content
    --textconv            for blob objects, run textconv on object's content
    --filters             for blob objects, run filters on object's content
    --path <blob>         use a specific path for --textconv/--filters
    --allow-unknown-type  allow -s and -t to work with broken/corrupt objects
    --buffer              buffer --batch output
    --batch[=<format>]    show info and content of objects fed from the standard input
    --batch-check[=<format>]
                          show info about obje

: 129

In [17]:
git cat-file -t 34fa038544bcd9aed660c08320214bafff94150b

tree


In [18]:
git cat-file -s 34fa038544bcd9aed660c08320214bafff94150b

202
